<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Построение-модели-для-классификации,-которая-выберет-подходящий-тариф" data-toc-modified-id="Построение-модели-для-классификации,-которая-выберет-подходящий-тариф-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Построение модели для классификации, которая выберет подходящий тариф</a></span><ul class="toc-item"><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span></li><li><span><a href="#Исследование-моделей" data-toc-modified-id="Исследование-моделей-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Исследование моделей</a></span><ul class="toc-item"><li><span><a href="#Decision-Tree-(Дерево-решений)" data-toc-modified-id="Decision-Tree-(Дерево-решений)-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Decision Tree (Дерево решений)</a></span></li><li><span><a href="#Random-Forest-(Случайный-лес)" data-toc-modified-id="Random-Forest-(Случайный-лес)-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Random Forest (Случайный лес)</a></span></li><li><span><a href="#Logistic-Regression-(Логистическая-регрессия)" data-toc-modified-id="Logistic-Regression-(Логистическая-регрессия)-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Logistic Regression (Логистическая регрессия)</a></span></li></ul></li><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Тестирование</a></span></li></ul></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Рекомендация тарифов для клиентов телеком компании

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [1]:
# загрузка библиотек
import pandas as pd

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier 
from sklearn.linear_model import LogisticRegression 

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.dummy import DummyClassifier

In [2]:
# чтение файла с данными и сохранение в переменную df
try:
    df = pd.read_csv('C:/Users/79853/Documents/datascience/yandex/Спринт7/users_behavior.csv')
except:
    df = pd.read_csv('/datasets/users_behavior.csv') # загрузка онлайн

In [3]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [4]:
# получение первых 5 строк таблицы df
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [5]:
# описание переменных
df.describe().T

count          mean          std  min         25%        50%  \
calls     3214.0     63.038892    33.236368  0.0     40.0000     62.000   
minutes   3214.0    438.208787   234.569872  0.0    274.5750    430.600   
messages  3214.0     38.281269    36.148326  0.0      9.0000     30.000   
mb_used   3214.0  17207.673836  7570.968246  0.0  12491.9025  16943.235   
is_ultra  3214.0      0.306472     0.461100  0.0      0.0000      0.000   

                 75%       max  
calls        82.0000    244.00  
minutes     571.9275   1632.06  
messages     57.0000    224.00  
mb_used   21424.7000  49745.73  
is_ultra      1.0000      1.00

### Вывод

- Целевая переменная `is_ultra`, которую необходимо предсказать, представлена в формате int со значениями 1/0. Для дальнейшего исследования такие значения подходят. 
- Пропуски в данных отсутствуют. 
- Переменные `calls` (количество звонков) и `messages` (количество sms-сообщений) предполагают целочисленные значения, их тип можно изменить на **int**.

## Предобработка данных

In [6]:
# изменение типа данных c float64 на int64 
df['messages'] = df['messages'].astype('int64')
df['calls'] = df['calls'].astype('int64')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   int64  
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   int64  
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 125.7 KB


In [8]:
# подсчёт явных дубликатов
df.duplicated().sum()

0

### Вывод

- Изменены типы данных переменных `calls` и `messages` на int.
- Явных дубликатов не выявлено

## Построение модели для классификации, которая выберет подходящий тариф

Поставленная задача относится к классу **«обучение с учителем»**, т.к. мы имеем набор данных и целевой признак, который нужно предсказать по остальным признакам, — **тариф пользователя** (Смарт или Ультра).

Целевой признак **категориальный**, т.е. необходимо решить задачу классификации. Категорий всего две — это **бинарная (двоичная) классификация**.

Для решения задачи классификации будем использовать следующие алгоритмы для определения лучшего:
1. Decision Tree (Дерево решений),
2. Random Forest (Случайный лес),
3. Logistic Regression (Логистическая регрессия),

Для проверки модели на правильность предсказаний нужно разбить данные на три части: обучающую, валидационную и тестовую. Размеры тестового и валидационного наборов обычно равны. Разобьем данные в соотношении 3:1:1.

Процентное соотношение от исходного датасета:
- обучающая выборка - 60%,
- валидационная выборка - 20%,
- тестовая выборка - 20%.

### Разбейте данные на выборки

In [9]:
# Разделим датасет на обучающую и валидационную выборки в соотношении 0,6 и 0,4
df_train, df_valid = train_test_split(df, test_size=0.4, random_state=123) 

In [10]:
# Разделим валидационную выборку пополам на валидационную и тестовую выборки
df_valid, df_test = train_test_split(df_valid, test_size=0.5, random_state=123) 

In [11]:
# посмотрим на размер сформированных выборок
print("обучающая выборка - {:.0%};".format(df_train.shape[0]/df.shape[0]), 'размер:', df_train.shape)
print("валидационная выборка - {:.0%};".format(df_valid.shape[0]/df.shape[0]), 'размер:', df_valid.shape)
print("тестовая выборка - {:.0%};".format(df_test.shape[0]/df.shape[0]), 'размер:', df_test.shape)

обучающая выборка - 60%; размер: (1928, 5)
валидационная выборка - 20%; размер: (643, 5)
тестовая выборка - 20%; размер: (643, 5)


In [12]:
# разделим выборки на признаки (features), по которым будут сделаны предсказания,
# и целевой признак (target), который будет предсказан
train_features = df_train.drop(['is_ultra'], axis=1)
train_target = df_train['is_ultra']

valid_features = df_valid.drop(['is_ultra'], axis=1)
valid_target = df_valid['is_ultra']

test_features = df_test.drop(['is_ultra'], axis=1)
test_target = df_test['is_ultra']

### Исследование моделей

#### Decision Tree (Дерево решений)

In [13]:
# определим лучшую модель
best_model_dt = None
best_result, best_depth = 0, 0
for depth in range(1, 100, 2):
    model = DecisionTreeClassifier(random_state=123, max_depth=depth)  # модель решающее дерево
    model.fit(train_features, train_target)  # обучение модели
    predictions = model.predict(valid_features)  # предсказания модели
    result = accuracy_score(valid_target, predictions)  # качество модели
    if result > best_result:
        best_depth = depth
        best_model_dt = model
        best_result = result        
print("Accuracy лучшей модели на валидационной выборке:", best_result, 'max_depth:', best_depth)

Accuracy лучшей модели на валидационной выборке: 0.8227060653188181 max_depth: 5


#### Random Forest (Случайный лес)

In [14]:
# определим лучшую модель с гиперпараметром n_estimators
best_model_rf = None
best_result, best_est = 0, 0
for est in range(1, 100, 2):
    model = RandomForestClassifier(random_state=123, n_estimators=est)  # модель
    model.fit(train_features, train_target)  # обучение модели на тренировочной выборке
    result = model.score(valid_features, valid_target)  # качество модели на валидационной выборке
    if result > best_result:
        best_est = est         # лучшее значение n_estimators
        best_model_rf = model  # лучшая модель
        best_result = result   # лучшее значение метрики accuracy на валидационных данных
print("Accuracy наилучшей модели на валидационной выборке:", best_result, 'n_estimators:', best_est)

Accuracy наилучшей модели на валидационной выборке: 0.8289269051321928 n_estimators: 39


In [15]:
# определим лучшую модель с гиперпараметрами n_estimators и max_depth
best_model_rf_2 = None
best_result, best_est, best_depth = 0, 0, 0
for est in range(1, 50, 2):
    for depth in range(1, 20):
        model = RandomForestClassifier(random_state=123, n_estimators=est, max_depth=depth)  # модель
        model.fit(train_features, train_target)  # обучение модели на тренировочной выборке
        result = model.score(valid_features, valid_target)  # качество модели на валидационной выборке
        if result > best_result:
            best_depth = depth
            best_est = est 
            best_model_rf_2 = model  # сохраните наилучшую модель
            best_result = result     #  сохраните наилучшее значение метрики accuracy на валидационных данных

print("Accuracy наилучшей модели на валидационной выборке:", best_result,
      'n_estimators:', best_est, 'max_depth:', best_depth)

Accuracy наилучшей модели на валидационной выборке: 0.8351477449455676 n_estimators: 21 max_depth: 11


#### Logistic Regression (Логистическая регрессия)

In [16]:
model_lr = LogisticRegression(solver='lbfgs', random_state=123) # модель логистической регрессии
model_lr.fit(train_features, train_target) # обучение модели на тренировочной выборке
result = model.score(valid_features, valid_target) # качество модели на валидационной выборке
print("Accuracy модели логистической регрессии на валидационной выборке:", result)

Accuracy модели логистической регрессии на валидационной выборке: 0.8102643856920684


### Тестирование

Протестируем модель с лучшим показателем метрики Accuracy на валидационной выборке:
- **best_model_rf_2 = RandomForestClassifier(random_state=123, n_estimators=21, max_depth=11)**

In [17]:
# тестирование модели Random Forest (Случайный лес) с двумя гиперпараметрами: n_estimators и max_depth
test_predictions_rf_2 = best_model_rf_2.predict(test_features)
accuracy_test_rf_2 = accuracy_score(test_target, test_predictions_rf_2)
print("Accuracy на тестовой выборке:", accuracy_test_rf_2)

Accuracy на тестовой выборке: 0.7978227060653188


## (бонус) Проверьте модели на адекватность

Для проверки модели на вменяемость применим **DummyClassifier** - классификатор, который делает прогнозы, используя простые правила. Этот классификатор служит простой базой для сравнения с другими более сложными классификаторами.

In [18]:
dummy_model = DummyClassifier(strategy='most_frequent',random_state=123)
dummy_model.fit(train_features, train_target)
predict_dummy = dummy_model.predict(test_features)
print('Accuracy модели DummyClassifier на тестовой выборке:', accuracy_score(test_target, predict_dummy))

Accuracy модели DummyClassifier на тестовой выборке: 0.6936236391912908


Простая модель DummyClassifier имеет Accuracy = 0.6936, что меньше Accuracy выбранной модели Random Forest (Случайный лес). Можно сделать вывод, что выбранная модель Random Forest (Случайный лес) является адекватной.

## Общий вывод

В ходе проекта были исследованы данные о поведении клиентов оператора мобильной связи, которые являются пользователями двух тарифных планов: «Смарт» и «Ультра». 

На **этапе предобработки данных** были изменены типы данных переменных `calls` и `messages` на int.

Далее были построены **модели для классификации**, которые предсказывают подходящий тариф для пользователя - «Смарт» или «Ультра». Для этого данные были разбиты на три выборки: **обучающую, валидационную и тестовую** в соотношении **3:1:1**. Были использованы следующие алгоритмы для построения моделей:
1. Decision Tree (Дерево решений),
2. Random Forest (Случайный лес) с одним гиперпараметром (n_estimators) и с двумя гиперпараметрами (n_estimators и max_depth),
3. Logistic Regression (Логистическая регрессия).

Лучшие показатели метрики Accuracy = **0.835** на валидационной выборке показала модель **RandomForestClassifier(random_state=123, n_estimators=21, max_depth=11)**. На тестовой выборке данная модель показала результаты Accuracy = **0.798**.

Дополнительно была проведена проверка модели на адекватность с помощью классификатора **DummyClassifier**, который делает прогнозы, используя простые правила. Значение Accuracy простой модели оказалось меньше Accuracy выбранной модели Random Forest, что позволило сделать вывод об адекватности выбранной модели.